eth/catalyst/simulated_beacon.go

```diff
@@ -0,0 +1,249 @@
+	lastBlockTime      uint64
+}
+
+func NewSimulatedBeacon(eth *eth.Ethereum) (*SimulatedBeacon, error) {
+	chainConfig := eth.APIBackend.ChainConfig()
+	if chainConfig.Dev == nil {
+		return nil, errors.New("incompatible pre-existing chain configuration")
+	}
+
+	return &SimulatedBeacon{
+		eth:           eth,
+		period:        chainConfig.Dev.Period,
+		shutdownCh:    make(chan struct{}),
+		buildWaitTime: time.Millisecond * 100,
+	}, nil
+}
+
+func (c *SimulatedBeacon) setFeeRecipient(feeRecipient common.Address) {
+	c.mu.Lock()
+	c.feeRecipient = feeRecipient
+	c.mu.Unlock()
+}
+
+// Start invokes the SimulatedBeacon life-cycle function in a goroutine
+func (c *SimulatedBeacon) Start() error {
+	go c.loop()
+	return nil
+}
+
+// Stop halts the SimulatedBeacon service
+func (c *SimulatedBeacon) Stop() error {
+	close(c.shutdownCh)
+	return nil
+}
+
+// beginSealing instructs the client to begin building a payload.
+func (c *SimulatedBeacon) beginSealing() (engine.ForkChoiceResponse, error) {
+	tstamp := uint64(time.Now().Unix())
+	if tstamp <= c.lastBlockTime {
+		tstamp = c.lastBlockTime + 1
+	}
+	c.mu.Lock()
+	feeRecipient := c.feeRecipient
+	c.mu.Unlock()
+
+	return c.engineAPI.ForkchoiceUpdatedV2(c.curForkchoiceState, &engine.PayloadAttributes{
+		Timestamp:             tstamp,
+		SuggestedFeeRecipient: feeRecipient,
+		Withdrawals:           c.withdrawals.queued(),
+	})
+}
+
+// finalizeSealing retrieves a completed payload and marks it as canonical if it contains transactions or withdrawals.
+func (c *SimulatedBeacon) finalizeSealing(id *engine.PayloadID, onDemand bool) error {
+	payload, err := c.engineAPI.GetPayloadV1(*id)
+	if err != nil {
+		return fmt.Errorf("error retrieving payload: %v", err)
+	}
+
+	if onDemand && len(payload.Transactions) == 0 && len(payload.Withdrawals) == 0 {
+		// If the payload is empty, despite there being pending transactions,
+		// that indicates that we need to give it more time to build the block.
+		if c.buildWaitTime < 10*time.Second {
+			c.buildWaitTime += c.buildWaitTime
+		}
+		return nil
+	}
+	c.buildWaitTime = 100 * time.Millisecond // Reset it
+	// mark the payload as canon
+	if _, err = c.engineAPI.NewPayloadV2(*payload); err != nil {
+		return fmt.Errorf("failed to mark payload as canonical: %v", err)
+	}
+	c.curForkchoiceState = engine.ForkchoiceStateV1{
+		HeadBlockHash:      payload.BlockHash,
+		SafeBlockHash:      payload.BlockHash,
+		FinalizedBlockHash: payload.BlockHash,
+	}
+	// mark the block containing the payload as canonical
+	if _, err = c.engineAPI.ForkchoiceUpdatedV2(c.curForkchoiceState, nil); err != nil {
+		return fmt.Errorf("failed to mark block as canonical: %v", err)
+	}
+	c.withdrawals.clearQueued()
+	c.lastBlockTime = payload.Timestamp
+	return nil
+}
+
+// loop manages the lifecycle of the SimulatedBeacon.
+// it drives block production, taking the role of a CL client and interacting with Geth via public engine/eth APIs
+func (c *SimulatedBeacon) loop() {
```

Comment:
I think we can split this big loop into two parts: on demand or regular, and also some fields can be initialized in the constructor. Please take a look.

```diff
diff --git a/eth/catalyst/simulated_beacon.go b/eth/catalyst/simulated_beacon.go
index b10ce497de..3e4a58657c 100644
--- a/eth/catalyst/simulated_beacon.go
+++ b/eth/catalyst/simulated_beacon.go
@@ -24,6 +24,7 @@ import (
 
 	"github.com/ethereum/go-ethereum/beacon/engine"
 	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core"
 	"github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/eth"
 	"github.com/ethereum/go-ethereum/log"
@@ -79,13 +80,14 @@ func (w *withdrawalQueue) clearQueued() {
 }
 
 type SimulatedBeacon struct {
-	shutdownCh   chan struct{}
-	eth          *eth.Ethereum
-	period       uint64
-	withdrawals  withdrawalQueue
-	feeRecipient common.Address
-	// mu gates concurrent access to the feeRecipient
-	mu                 sync.Mutex
+	shutdownCh  chan struct{}
+	eth         *eth.Ethereum
+	period      uint64
+	withdrawals withdrawalQueue
+
+	feeRecipient     common.Address
+	feeRecipientLock sync.Mutex // lock gates concurrent access to the feeRecipient
+
 	engineAPI          *ConsensusAPI
 	curForkchoiceState engine.ForkchoiceStateV1
 	buildWaitTime      time.Duration
@@ -97,24 +99,44 @@ func NewSimulatedBeacon(eth *eth.Ethereum) (*SimulatedBeacon, error) {
 	if chainConfig.Dev == nil {
 		return nil, errors.New("incompatible pre-existing chain configuration")
 	}
+	header := eth.BlockChain().CurrentHeader()
+	current := engine.ForkchoiceStateV1{
+		HeadBlockHash:      header.Hash(),
+		SafeBlockHash:      header.Hash(),
+		FinalizedBlockHash: header.Hash(),
+	}
+	engineAPI := NewConsensusAPI(eth)
 
+	// if genesis block, send forkchoiceUpdated to trigger transition to PoS
+	if header.Number.Sign() == 0 {
+		if _, err := engineAPI.ForkchoiceUpdatedV2(current, nil); err != nil {
+			return nil, err
+		}
+	}
 	return &SimulatedBeacon{
-		eth:           eth,
-		period:        chainConfig.Dev.Period,
-		shutdownCh:    make(chan struct{}),
-		buildWaitTime: time.Millisecond * 100,
+		eth:                eth,
+		period:             chainConfig.Dev.Period,
+		shutdownCh:         make(chan struct{}),
+		buildWaitTime:      time.Millisecond * 100,
+		engineAPI:          engineAPI,
+		lastBlockTime:      header.Time,
+		curForkchoiceState: current,
 	}, nil
 }
 
 func (c *SimulatedBeacon) setFeeRecipient(feeRecipient common.Address) {
-	c.mu.Lock()
+	c.feeRecipientLock.Lock()
 	c.feeRecipient = feeRecipient
-	c.mu.Unlock()
+	c.feeRecipientLock.Unlock()
 }
 
 // Start invokes the SimulatedBeacon life-cycle function in a goroutine
 func (c *SimulatedBeacon) Start() error {
-	go c.loop()
+	if c.period == 0 {
+		go c.loopOnDemand()
+	} else {
+		go c.loop()
+	}
 	return nil
 }
 
@@ -130,9 +152,9 @@ func (c *SimulatedBeacon) beginSealing() (engine.ForkChoiceResponse, error) {
 	if tstamp <= c.lastBlockTime {
 		tstamp = c.lastBlockTime + 1
 	}
-	c.mu.Lock()
+	c.feeRecipientLock.Lock()
 	feeRecipient := c.feeRecipient
-	c.mu.Unlock()
+	c.feeRecipientLock.Unlock()
 
 	return c.engineAPI.ForkchoiceUpdatedV2(c.curForkchoiceState, &engine.PayloadAttributes{
 		Timestamp:             tstamp,
@@ -147,7 +169,6 @@ func (c *SimulatedBeacon) finalizeSealing(id *engine.PayloadID, onDemand bool) e
 	if err != nil {
 		return fmt.Errorf("error retrieving payload: %v", err)
 	}
-
 	if onDemand && len(payload.Transactions) == 0 && len(payload.Withdrawals) == 0 {
 		// If the payload is empty, despite there being pending transactions,
 		// that indicates that we need to give it more time to build the block.
@@ -175,65 +196,53 @@ func (c *SimulatedBeacon) finalizeSealing(id *engine.PayloadID, onDemand bool) e
 	return nil
 }
 
-// loop manages the lifecycle of the SimulatedBeacon.
-// it drives block production, taking the role of a CL client and interacting with Geth via public engine/eth APIs
-func (c *SimulatedBeacon) loop() {
+func (c *SimulatedBeacon) loopOnDemand() {
 	var (
-		ticker = time.NewTimer(time.Second * time.Duration(c.period))
-		header = c.eth.BlockChain().CurrentHeader()
+		newTxs = make(chan core.NewTxsEvent)
+		sub    = c.eth.TxPool().SubscribeNewTxsEvent(newTxs)
 	)
+	defer sub.Unsubscribe()
 
-	c.curForkchoiceState = engine.ForkchoiceStateV1{
-		HeadBlockHash:      header.Hash(),
-		SafeBlockHash:      header.Hash(),
-		FinalizedBlockHash: header.Hash(),
-	}
-	c.engineAPI = NewConsensusAPI(c.eth)
-	c.lastBlockTime = header.Time
-
-	// if genesis block, send forkchoiceUpdated to trigger transition to PoS
-	if header.Number.Sign() == 0 {
-		if _, err := c.engineAPI.ForkchoiceUpdatedV2(c.curForkchoiceState, nil); err != nil {
-			log.Error("failed to initiate PoS transition for genesis via Forkchoiceupdated", "err", err)
+	for {
+		select {
+		case <-c.shutdownCh:
 			return
+		case <-newTxs:
+			resp, err := c.beginSealing()
+			if err != nil {
+				log.Error("Error starting sealing-work", "err", err)
+				return
+			}
+			time.Sleep(c.buildWaitTime)
+
+			err = c.finalizeSealing(resp.PayloadID, true)
+			if err != nil {
+				log.Error("Error collecting sealing-work", "err", err)
+				return
+			}
 		}
 	}
+}
 
-	var fcId *engine.PayloadID
-	if fc, err := c.beginSealing(); err != nil {
-		log.Error("Error starting sealing-work", "err", err)
-		return
-	} else {
-		fcId = fc.PayloadID
-	}
-	onDemand := (c.period == 0)
+// loop manages the lifecycle of the SimulatedBeacon.
+// it drives block production, taking the role of a CL client and interacting with Geth via public engine/eth APIs
+func (c *SimulatedBeacon) loop() {
+	timer := time.NewTimer(time.Second * time.Duration(c.period))
 	for {
+		resp, err := c.beginSealing()
+		if err != nil {
+			log.Error("Error starting sealing-work", "err", err)
+			return
+		}
 		select {
 		case <-c.shutdownCh:
 			return
-		case <-ticker.C:
-			if onDemand {
-				// Do nothing as long as blocks are empty
-				if pendingTxs, _ := c.eth.APIBackend.TxPool().Stats(); pendingTxs == 0 && len(c.withdrawals.queued()) == 0 {
-					ticker.Reset(c.buildWaitTime)
-					continue
-				}
-			}
-			if err := c.finalizeSealing(fcId, onDemand); err != nil {
+		case <-timer.C:
+			if err := c.finalizeSealing(resp.PayloadID, false); err != nil {
 				log.Error("Error collecting sealing-work", "err", err)
 				return
 			}
-			if fc, err := c.beginSealing(); err != nil {
-				log.Error("Error starting sealing-work", "err", err)
-				return
-			} else {
-				fcId = fc.PayloadID
-			}
-			if !onDemand {
-				ticker.Reset(time.Second * time.Duration(c.period))
-			} else {
-				ticker.Reset(c.buildWaitTime)
-			}
+			timer.Reset(time.Second * time.Duration(c.period))
 		}
 	}
 }

```

Reply:
One issue with this is that if we are in on-demand mode and the payload takes longer to build than `buildWaitTime`, the block will not be built.

Reply:
Yup, but you have the mechanism to double the wait time. So the wait time will gradually climb to a proper value, isn't it?

This logic is not changed.

Reply:
I should have elaborated...  the issue is that in `loopOnDemand` the build is triggered by the `NewTxs` event.  So if a block takes too long to build, the build won't re-trigger until there are new txs in the pool.

Reply:
We can use a new API to retrieve payload in sync mode. https://github.com/rjl493456442/go-ethereum/commit/9f467bccd130d77efe2a48026f64930cfe5c0f5e

But yeah, if something weird happens(we have an executable transaction but later on it's gone somehow) then GetFullPayload will still return empty-one, but I think it's fine to dev mode.

Reply:
If we receive a new transaction in the pool, the build triggers, and the transaction takes longer to build than the default time, the build loop won't re-trigger until there are other transactions in the pool.

Reply:
No, you don't need the wait mechanism if you get payload in blocked way.